### PR TITLE
add pickMultipleRandom function

### PR DESCRIPTION
**Pull request by woylie**
I added a pickMultipleRandom function modeled after pickRandom. I already wrote additional functions for weighted random picking, but since this is my first attempt at contributing to an open source project, I'd rather like to listen to your comments before putting it all out at once.

### CODE REVIEW

**Comment by josdejong** [comment]
Thanks for your PR @woylie, and sorry for my late response. Your code look neat and well tested, and it's a usable function.

I'm wondering though whether we compose this functionality by combining two more generic functions, something with a `repeat` function and `pickRandom`, but I can't come up with a user friendly solution for that.

File: lib/expression/docs/function/probability/pickMultipleRandom.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
+module.exports = {
+  'name': 'pickMultipleRandom',
+  'category': 'Probability',
+  'syntax': [
+    'pickMultipleRandom(array, number)'
+  ],
+  'description':
+      'Pick a given number of random entries from a given array.',
+  'examples': [
+    'pickRandom([1, 3, 1, 6], 2)'
```
Comment (josdejong) [inline lib/expression/docs/function/probability/pickMultipleRandom.js]: I think the example should call `pickMultipleRandom` instead of `pickRandom`

**Comment by woylie** [comment]
The problem with using pickRandom from a repeat function would be that all the validations would be executed again with each pick. That's why I wrote a separate function.

I wrote two more functions called pickWeightedRandom and pickMultipleRandomWeighted. Having four different functions for nearly the same purpose seems a bit unnecessary to me. I'd suggest to rewrite pickRandom to include all of the functions:

pickRandom(possibles, number, weights)

possibles: array of values to pick from, as before
number: number of values to pick, optional, default 1
weight: array of weights with the same length as possibles, optional

To me this seems to be the cleanest way to extend the functionality without breaking legacy code. What do you think? Give me a go and I'll get on it.

**Comment by josdejong** [comment]
Makes a lot of sense to extend the `pickRandom` function with extra options instead of creating more functions!

Possible options:
```js
pickRandom(possibles, number, weights)   // your suggestion
pickRandom(possibles, weights, number)
pickRandom(possibles, {weights: [...], number: number})   // object with extra options. more cumbersome
```

I like your suggestion most. It would be neat if you could call the function in all possible combinations:

```js
pickRandom(possibles)
pickRandom(possibles, number)
pickRandom(possibles, number, weights)
pickRandom(possibles, weights)
```
since weights is an array and number a scalar value, this should be doable. Would be great if you could give this a try! 

The file distribution.js can use some love, it's more complicated than necessary, and it still has to be rewritten into a `typed-function`. Feel free to do a refactoring there.

**Comment by woylie** [comment]
Good idea to enable all combinations. I'm going to have a look at your typed-functions as well.

**Comment by woylie** [comment]
If number > 1, the return value must be an array, yet currently pickRandom returns a single value. I think it would be inconsistent to return a single value for number == 1 and an array for number == 0 and number > 1. But of course it would be a breaking change if now pickRandom would now return an array even if no number argument was passed. It would of course be possible to return a single value if the number argument wasn't passed and an array if it was, but while this would not break old code, this behavior would seem odd and confusing to me. What do you think about this?

**Comment by josdejong** [comment]
Good point. From a puristic point of view it sounds inconsistent but it's really the only practical way. You don't want `pickRandom(array)` to return an array containing a single value or something like that. I think it's OK when `pickRandom` returns either a value or an array depending on the input arguments.

**Comment by woylie** [comment]
I've rewritten the pickRandom function now as discussed, but it still needs some refactoring. But I'm also wondering how to make it work within the context of distribution.js. If no weights are passed, it could just make use of the uniform or normal distribution. But if weights are passed, that doesn't seem to make much sense to me. Any thoughts on this? Did you have any plans or ideas for a new version of distribution.js?

**Comment by josdejong** [comment]
> Did you have any plans or ideas for a new version of distribution.js?

The file can use a big refactoring. If you're interested in picking this up you're welcome! Just start clean and rethink what functions and syntax we need exactly (preferably without breaking changes). Strip out all unused logic, make it simpler, and just keep the stuff that we actually use, refactor the functions into typed-functions so we get a more neat structure. Something like that. 

I have to give the behavior regarding uniform/normal/weighted a bit more thought. Feel free to share your ideas in this regard.

**Comment by woylie** [comment]
I'm not sure whether it would make sense to include the weighted distribution in the available distributions. The weights array only relates to the given possibles array and can't be reused, in contrast to uniform and normal distributions. As soon as a weights array is passed, we just have to ignore the distribution of the distribution object. The option to pass weights doesn't really make sense inside a distribution object.

I'm also wondering whether it would be a good idea to rewrite distribution.js so that the parameters (mean and sd for normal distribution) can be passed as well. This might be handy if we're going to add more distributions or statistical functions.

**Comment by woylie** [comment]
Rewriting everything as a typed function when no optional arguments are allowed is less than ideal. While it is nice not having to worry about validating the arguments, you'll always end up with multiple signatures that share the main part of the code, which then forces you to write additional helper functions to avoid code duplication. And then the typed function ends up only sorting out the arguments. Also it seems like it is not possible to add a signature with no arguments, is that right?

**Comment by josdejong** [comment]
I will give this some more thought coming weekend (when I'm back from vacation).

As for the typed-functions: it indeed lacks optional arguments forcing you to write a helper functions. I expect though that is may still save you code or at least is very little overhead (you still don't have to write type checks and error messages). If it actually gives much overhead it's ok with me not to use typed-functions. They should help you, not get in your way.

You can indeed write typed functions without signature, like 

```js
var fn = math.typed({
  '': function () {return 42}
})
```

**Comment by woylie** [comment]
As for the new pickRandom function, I'd leave it at that. I could extract some logic and put it into further functions, but I'm not sure that's necessary. It seems pretty clear to me as it is, but if you have any suggestions, let me know.

As for the refactoring, I turned randomInt() into a typed function (see commits). However, I ran into problems when trying to turn random() into a typed function as well. Because I couldn't find the cause (the typed version of the random function was nearly identical to the working typed randomInt function), I copied the old random function and made a new typed random function that just forwards the arguments to the old function, just for testing:

```
      random: typed({
          '': function() {
            return _randomOld();
          },
          'Array | number': function(arg1) {
            return _randomOld(arg1);
          },
          'Array | number, number': function(arg1, arg2) {
            return _randomOld(arg1, arg2);
          },
          'Array, number, number': function(arg1, arg2, arg3) {
            return _randomOld(arg1, arg2, arg3);
          },
        }),
```

I expected this to work, because all I did was to insert an intermediary function without any additional logic. However, the tests fail when a matrix is used as the only argument. Apparently the function won't return a matrix anymore:

```
1) distribution random should pick uniformly distributed random matrix, with elements in [0, 1]:

      AssertionError: false == true
      + expected - actual

      -false
      +true
      
      at test/function/probability/distribution.test.js:141:9
      at Array.forEach (native)
      at Context.<anonymous> (test/function/probability/distribution.test.js:140:16)
```

Do you know why this could be?

Aside from this, I was thinking about creating a helper function for both random() and randomInt(), as these functions are nearly identical, except for the use of _random() and _randomInt() respectively, and the possibility of passing no arguments to _random(). Something like:

```
random: function(arg1, arg2, arg3) {
    _randomHelper(_random, arg1, arg2, arg3);
}

randomInt: function(arg1, arg2, arg3) {
    _randomHelper(_randomInt, arg1, arg2, arg3);
}
```

However, to write _randomHelper as a typed function, we cannot just pass (_random, arg1, arg2, arg3), because then the typed function would always evaluate four arguments, even if originally only 1 or 2 were passed. We can't use _randomHelper.apply(this, arguments), because we have to pass the random function as an additional argument. My understanding is that this would only work correctly with the ES6 spread operator: _randomHelper(_random, ...arguments), but as we don't use babel, this isn't an option. So unless you have a better idea on how to extract the common logic of both functions, we'll have to live with some duplication.

**Comment by josdejong** [comment]
Thanks, [your code](https://github.com/woylie/mathjs/blob/master/lib/function/probability/distribution.js) looks neat so far!

As for the error you get: I'm not sure from your description where this could originate. I tested whether it could be some bug in `typed-function` but that seems to work, see jsbin: http://jsbin.com/xotuxih/edit?html,output. I've run the tests of your clone of mathjs but these all succeed so I guess you didn't yet commit your work?

You indeed probably need a helper function. There are various ways to pass the arguments. One way is to do some simple checks inside the function whether `arg1`, `arg2`, and `arg3` are undefined. Or you could pass these arguments as an object like `_randomHelper(_random, {arg1: ..., arg2: ..., arg3: ...})`. In some cases it's possible to unify the function calls by choosing proper default values when some arguments are not passed by the user, like when the user calls `math.random()`, call an internal helper function like `_random(0 /*min*/, 1 /*max*/)`. It requires some fiddling around to find a neat solution I think.

**Comment by woylie** [comment]
Right, I hadn't committed this test. You can have a look at it in the latest commit.

As for the other issue: Yes, it is clear that I'll need a helper function. Perhaps I wasn't clear enough in my explanation.

```
random: function(arg1, arg2, arg3) {
    _randomHelper(_random, arg1, arg2, arg3);
}

var _randomHelper = math.typed({
    'function': function(randFunc) {/*...*/},
    'function, Array | number': function(/*...*/) {},
    'function, Array | number, number': function(/*...*/) {},
    'function, Array, number, number': function(/*...*/) {},
});

random(10);
// random will call _randomHelper(_random, 10, undefined, undefined), which doesn't fit any signature.
```

With ES6, the following would be possible. No undefined would be passed. Unfortunately not available to us right now.

```
random: function(arg1, arg2, arg3) {
    _randomHelper(_random, ...arguments);
}
```

Without ES6, we could do something like this. Not very elegant, but it works.

```
random: function(arg1, arg2, arg3) {
    switch (arguments.length) {
    case 0:
        return _randomHelper(_random);
    case 1:
        return _randomHelper(_random, arg1);
    case 2:
        return _randomHelper(_random, arg1, arg2);
    case 3:
        return _randomHelper(_random, arg1, arg2, arg3);
    default:
        throw new ArgumentsError('random', arguments.length, 0, 3);
    }
},
```

So we can use code duplication, switch, or a manual validation of the arguments without typed function. Turning it around and making random/randomInt into typed functions instead of _randomHelper wouldn't help us with the duplication. What do you think?

We could of course also define _randomHelper like this, but that would be very ugly:

```
var _randomHelper = math.typed({
    'function, undefined, undefined, undefined': function(randFunc) {/*...*/},
    /* etc. */
});
```

**Comment by woylie** [comment]
PS: I've made another commit with the switch solution. It fails the same way the commit before fails.

**Comment by woylie** [comment]
PPS: This happens when arg 1 is a matrix (only the essential parts):

```
random: function(arg1, arg2, arg3) {
    console.log(arg1 && arg1.isMatrix === true); // true
    return _randomHelper(_random, arg1);
}

var _randomHelper = typed({
    'function, number | Array': function(randFunc, arg) {
        console.log(arg && arg.isMatrix === true); // false
    },
}
```

**Comment by josdejong** [comment]
Sorry I wasn't able to reply earlier :( 

Using ES6 would be nice indeed but we're not using ES6 (yet) in the math.js code, so you will have to go with the less elegant structure.

About https://github.com/josdejong/mathjs/pull/679#issuecomment-236763513: That's correct. An Array is different from a Matrix. You can either make two implementations (one for Arrays, one for Matrices), or create a single Array implementation and convert Matrices to Arrays and back. The general behavior of math.js is to output an Array when input contains Array(s), and output a Matrix when inputs contain a Matrix.

I see there is only one unit test failing. The reason for this is that this unit test expects Matrices whilst it gets Arrays on this line: https://github.com/woylie/mathjs/blob/master/test/function/probability/distribution.test.js#L141

**Comment by woylie** [comment]
Has it really been two months? I just reverted the changes I made for demonstration purposes and made some minor adjustments. Typed functions don't seem to be the panacea here, so I'd leave it at that for now. The tests are passing, so if you don't have any urgent requests for this PR, feel free to merge. distributions.js could still need some attention, but right now I'm not sure how to approach it.

**Comment by josdejong** [comment]
Time flies :D

Thanks for updating the PR, looks good!

**Pull request merged**